modules/highlight/highlight.go

```diff
@@ -179,61 +177,90 @@ func File(numLines int, fileName, language string, code []byte) []string {
 
 	iterator, err := lexer.Tokenise(nil, string(code))
 	if err != nil {
-		log.Error("Can't tokenize code: %v", err)
-		return plainText(string(code), numLines)
+		return nil, fmt.Errorf("can't tokenize code: %w", err)
 	}
 
-	err = formatter.Format(htmlw, styles.GitHub, iterator)
+	err = formatter.Format(htmlWriter, styles.GitHub, iterator)
 	if err != nil {
-		log.Error("Can't format code: %v", err)
-		return plainText(string(code), numLines)
-	}
-
-	htmlw.Flush()
-	finalNewLine := false
-	if len(code) > 0 {
-		finalNewLine = code[len(code)-1] == '\n'
+		return nil, fmt.Errorf("can't format code: %w", err)
 	}
 
-	m := make([]string, 0, numLines)
-	for i, v := range strings.SplitN(htmlbuf.String(), "\n", numLines) {
-		content := string(v)
+	_ = htmlWriter.Flush()
 
-		// remove useless wrapper nodes that are always present
-		content = strings.Replace(content, "<span class=\"line\"><span class=\"cl\">", "", 1)
-		content = strings.TrimPrefix(content, `</span></span>`)
+	m := make([]string, 0, bytes.Count(code, []byte{'\n'})+1)
 
-		// if there's no final newline, closing tags will be on last line
-		if !finalNewLine && i == numLines-1 {
-			content = strings.TrimSuffix(content, `</span></span>`)
+	htmlStr := htmlBuf.String()
+	line := strings.Builder{}
+	insideLine := 0 // every <span class="cl"> makes it increase one level, every closed <span class="cl"> makes it decrease one level
+	tagStack := make([]string, 0, 4)
+	for len(htmlStr) > 0 {
+		pos1 := strings.IndexByte(htmlStr, '<')
+		pos2 := strings.IndexByte(htmlStr, '>')
```

Comment:
This will break when code contains `<` or `>`, e.g. `<span><</span>` is rendered as `<` by browsers.

Reply:
I am not sure what do you mean. The HTML string is generated by Chroma, which is strict and stable. I do not think `<span><</span>` is standard or it would be generated by Chroma.

If you worry about it, I will add a special case for `<`.

Reply:
Yes it seems Chroma does the escaping of `<`, so it's actually fine.